src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineWindowsLauncher.java

```diff
@@ -0,0 +1,269 @@
+        return conn.exec(checkCommand, logger) == 0;
+
+    }
+
+    private boolean authenticateSSH(WindowsConfiguration windowsConfig, Connection sshConnection,
+                                    TaskListener listener) throws Exception{
+        boolean isAuthenticated;
+        String windowsUsername = windowsConfig.getWindowsUsername();
+        if (windowsConfig.getPrivateKeyCredentialsId() != null) {
+            StandardUsernameCredentials cred = windowsConfig.getPrivateKeyCredentials();
+            isAuthenticated = SSHAuthenticator.newInstance(sshConnection, cred, windowsUsername).authenticate(listener);
+        } else {
+            isAuthenticated = sshConnection.authenticateWithPassword(windowsUsername, windowsConfig.getPassword());
+        }
+        return isAuthenticated;
+    }
+
+    private boolean bootstrap(ComputeEngineComputer computer, TaskListener listener) throws IOException,
+            Exception { //TODO: better exceptions
+        logInfo(computer, listener, "bootstrap");
+        ComputeEngineInstance node = computer.getNode();
+        WindowsConfiguration windowsConfig = node.windowsConfig;
+        if (node == null) {
+            throw new IllegalArgumentException("A ComputeEngineComputer with no node was provided");
+        }
+        Connection bootstrapConn = null;
+        try {
+            int tries = bootstrapAuthTries;
+            boolean isAuthenticated = false;
+            while (tries-- > 0) {
+                logInfo(computer, listener, "Authenticating as " + windowsConfig.getWindowsUsername());
+                try {
+                    bootstrapConn = connectToSsh(computer, listener);
+                    isAuthenticated = authenticateSSH(windowsConfig, bootstrapConn, listener);
+                } catch (IOException e) {
+                    logException(computer, listener, "Exception trying to authenticate", e);
+                    bootstrapConn.close();
+                }
+                if (isAuthenticated) {
+                    break;
+                }
+                logWarning(computer, listener, "Authentication failed. Trying again...");
+                Thread.sleep(bootstrapAuthSleepMs);
+            }
+            if (!isAuthenticated) {
+                logWarning(computer, listener, "Authentication failed");
+                return false;
+            }
+        } finally {
+            if (bootstrapConn != null) {
+                bootstrapConn.close();
+            }
+        }
+        return true;
+    }
+
+    private Connection connectToSsh(ComputeEngineComputer computer, TaskListener listener) throws Exception {
+        ComputeEngineInstance node = computer.getNode();
+        if (node == null) {
+            throw new IllegalArgumentException("A ComputeEngineComputer with no node was provided");
+        }
+
+        final long timeout = node.getLaunchTimeoutMillis();
+        final long startTime = System.currentTimeMillis();
+        while (true) {
```

Comment:
while (true) is very code smelly. Consider an using ExponentialBackoff strategy here as well.